pkg/reconciler/route/traffic/rollout.go

```diff
@@ -70,3 +73,100 @@ type RevisionRollout struct {
 	// target percentage.
 	Percent int `json:"percent"`
 }
+
+// Step processes previous Rollout object and compares to the current
+// rollout state. If there is different, Step will start or stop the rollout
+// and updates `RevisionRollout` objects of `cur` accoringly.
+// At the end of the call `cur` contains the desired traffic shape.
+// Step returns `true` if any changes have been made.
+func (cur *Rollout) Step(prev *Rollout) bool {
+	if prev == nil {
+		return false
+	}
+
+	// The algorithm below is simplest, but probably not the most performant.
+	// Optimize in the later passes.
+	// Map by tag.
+	currConfigs, prevConfigs := map[string][]*ConfigurationRollout{}, map[string][]*ConfigurationRollout{}
+	for i, cfg := range cur.Configurations {
+		currConfigs[cfg.Tag] = append(currConfigs[cfg.Tag], &cur.Configurations[i])
+	}
+	for i, cfg := range prev.Configurations {
+		prevConfigs[cfg.Tag] = append(prevConfigs[cfg.Tag], &prev.Configurations[i])
+	}
+	ret := false
+	for t, ccfgs := range currConfigs {
+		pcfgs, ok := prevConfigs[t]
```

Comment:
ccfgs = tagGoal (the goal state of this tag once the rollout is complete)
pcfgs = tagState (the state of this tag right now)